File: packages/python/src/kingpepe_sdk/payments.py
# Copyright (c) 2026 The KingPepe developers
# Distributed under the MIT software license, see the accompanying LICENSE file.
"""Merchant payment building blocks (pure state machine + RPC observer).

This is NOT a hosted payment processor. Assign a UNIQUE address per payment,
persist every request and its last state, treat webhook events as at-least-once
(idempotent on ``id``), and never trust a payment until it reaches the configured
confirmation depth. Times are Unix milliseconds for cross-language consistency.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Any, Literal

from .address import assert_kingpepe_address
from .errors import ValidationError
from .money import rpc_amount_to_base_units
from .uri import KingPepeUri, build_uri

PaymentStatus = Literal[
    "pending", "detected", "confirming", "paid", "underpaid", "overpaid", "expired", "cancelled", "failed"
]

_TERMINAL = {"paid", "overpaid", "expired", "cancelled", "failed"}


def _now_ms() -> int:
    return int(time.time() * 1000)


@dataclass
class PaymentRequest:
    id: str
    address: str
    amount_base_units: int
    required_confirmations: int = 1
    expires_at: int = 0
    label: str | None = None
    message: str | None = None


@dataclass
class PaymentObservation:
    received_base_units: int = 0
    confirmed_base_units: int = 0
    confirmations: int = 0
    seen: bool = False


@dataclass
class PaymentState:
    status: PaymentStatus
    request: PaymentRequest
    observation: PaymentObservation
    terminal: bool


def create_payment_request(
    *,
    id: str,
    address: str,
    amount_base_units: int,
    required_confirmations: int = 1,
    ttl_ms: int = 15 * 60 * 1000,
    label: str | None = None,
    message: str | None = None,
    now: int | None = None,
) -> PaymentRequest:
    if not id:
        raise ValidationError("Payment id is required (idempotency key).")
    assert_kingpepe_address(address)
    if amount_base_units <= 0:
        raise ValidationError("Payment amount must be positive.")
    if required_confirmations < 0:
        raise ValidationError("required_confirmations must be >= 0.")
    started = now if now is not None else _now_ms()
    return PaymentRequest(
        id=id,
        address=address,
        amount_base_units=amount_base_units,
        required_confirmations=required_confirmations,
        expires_at=started + ttl_ms,
        label=label,
        message=message,
    )


def payment_uri(request: PaymentRequest) -> str:
    return build_uri(
        KingPepeUri(
            address=request.address,
            amount_base_units=request.amount_base_units,
            label=request.label,
            message=request.message,
        )
    )


def evaluate_payment(
    request: PaymentRequest, observation: PaymentObservation, now: int | None = None
) -> PaymentState:
    """Pure payment-state evaluator. Deterministic; no I/O."""
    moment = now if now is not None else _now_ms()

    def state(status: PaymentStatus) -> PaymentState:
        return PaymentState(status=status, request=request, observation=observation, terminal=status in _TERMINAL)

    expected = request.amount_base_units
    confirmed = observation.confirmed_base_units
    received = observation.received_base_units

    if confirmed >= expected:
        return state("overpaid" if confirmed > expected else "paid")
    if moment >= request.expires_at:
        return state("underpaid" if confirmed > 0 else "expired")
    if not observation.seen or received == 0:
        return state("pending")
    if observation.confirmations < request.required_confirmations:
        return state("detected" if observation.confirmations == 0 else "confirming")
    return state("underpaid")


def to_payment_event(state: PaymentState, now: int | None = None) -> dict[str, Any]:
    """Build a webhook-ready, JSON-serializable event object."""
    return {
        "type": "payment_status",
        "payment_id": state.request.id,
        "status": state.status,
        "address": state.request.address,
        "expected_base_units": str(state.request.amount_base_units),
        "received_base_units": str(state.observation.received_base_units),
        "confirmations": state.observation.confirmations,
        "required_confirmations": state.request.required_confirmations,
        "at": now if now is not None else _now_ms(),
    }


def observe_payment(client: Any, request: PaymentRequest) -> PaymentObservation:
    """Observe a payment via a sync client's ``list_unspent`` for the address.

    Assumes a unique, unspent-until-settled address. For swept funds, track
    receipts in your own database instead.
    """
    utxos = client.list_unspent(0, 9_999_999, [request.address])
    received = 0
    confirmed = 0
    max_conf = 0
    seen = False
    for u in utxos:
        seen = True
        base = rpc_amount_to_base_units(u["amount"])
        received += base
        if u["confirmations"] >= request.required_confirmations:
            confirmed += base
        max_conf = max(max_conf, u["confirmations"])
    return PaymentObservation(
        received_base_units=received, confirmed_base_units=confirmed, confirmations=max_conf, seen=seen
    )


def check_payment(client: Any, request: PaymentRequest, now: int | None = None) -> PaymentState:
    """Observe + evaluate with a sync client."""
    return evaluate_payment(request, observe_payment(client, request), now)

File: packages/javascript/src/payments.ts
// Copyright (c) 2026 The KingPepe developers
// Distributed under the MIT software license, see the accompanying LICENSE file.

import { KingPepeClient } from "./client.js";
import { assertKingPepeAddress } from "./address.js";
import { rpcAmountToBaseUnits } from "./money.js";
import { buildUri } from "./uri.js";
import { ValidationError } from "./errors.js";

/**
 * Merchant payment building blocks. This is NOT a hosted payment processor: it
 * provides reusable, side-effect-light pieces (a pure state evaluator plus an
 * RPC-backed observer) that a merchant backend composes with its own database.
 *
 * Production guidance: assign a UNIQUE address per payment, persist every
 * PaymentRequest and its last PaymentState, treat webhook events as at-least-once
 * (idempotent on `request.id`), and never trust a payment until it reaches the
 * configured confirmation depth.
 */

export type PaymentStatus =
  | "pending"
  | "detected"
  | "confirming"
  | "paid"
  | "underpaid"
  | "overpaid"
  | "expired"
  | "cancelled"
  | "failed";

export interface PaymentRequest {
  /** Merchant-assigned unique id (used as the idempotency key). */
  id: string;
  /** Unique receiving address for this payment. */
  address: string;
  /** Expected amount in integer base units (sat). */
  amountBaseUnits: bigint;
  requiredConfirmations: number;
  /** Absolute expiry as a Unix ms timestamp. */
  expiresAt: number;
  label?: string;
  message?: string;
}

export interface PaymentObservation {
  /** Total amount received to the address in base units (0-conf included). */
  receivedBaseUnits: bigint;
  /** Amount received with at least `requiredConfirmations` confirmations. */
  confirmedBaseUnits: bigint;
  /** Highest confirmation count observed among received outputs (0 if none). */
  confirmations: number;
  /** Whether any output (even 0-conf) has been seen. */
  seen: boolean;
}

export interface PaymentState {
  status: PaymentStatus;
  request: PaymentRequest;
  observation: PaymentObservation;
  /** True once the terminal outcome (paid/overpaid/underpaid/expired/...) is reached. */
  terminal: boolean;
}

/** A webhook-ready, serializable event describing a payment state. */
export interface PaymentEvent {
  type: "payment_status";
  paymentId: string;
  status: PaymentStatus;
  address: string;
  expectedBaseUnits: string;
  receivedBaseUnits: string;
  confirmations: number;
  requiredConfirmations: number;
  /** Unix ms timestamp when the event was produced. */
  at: number;
}

export interface CreatePaymentRequestInput {
  id: string;
  address: string;
  amountBaseUnits: bigint;
  requiredConfirmations?: number;
  /** Time-to-live in ms from `now`. Defaults to 15 minutes. */
  ttlMs?: number;
  label?: string;
  message?: string;
  now?: number;
}

const TERMINAL: ReadonlySet<PaymentStatus> = new Set([
  "paid",
  "overpaid",
  "expired",
  "cancelled",
  "failed",
]);

/** Create a validated PaymentRequest. */
export function createPaymentRequest(input: CreatePaymentRequestInput): PaymentRequest {
  if (!input.id) throw new ValidationError("Payment id is required (idempotency key).");
  assertKingPepeAddress(input.address);
  if (input.amountBaseUnits <= 0n) throw new ValidationError("Payment amount must be positive.");
  const requiredConfirmations = input.requiredConfirmations ?? 1;
  if (requiredConfirmations < 0) throw new ValidationError("requiredConfirmations must be >= 0.");
  const now = input.now ?? Date.now();
  const ttlMs = input.ttlMs ?? 15 * 60 * 1000;
  return {
    id: input.id,
    address: input.address,
    amountBaseUnits: input.amountBaseUnits,
    requiredConfirmations,
    expiresAt: now + ttlMs,
    label: input.label,
    message: input.message,
  };
}

/** Build the `kingpepe:` URI for a payment request. */
export function paymentUri(request: PaymentRequest): string {
  return buildUri({
    address: request.address,
    amountBaseUnits: request.amountBaseUnits,
    label: request.label,
    message: request.message,
  });
}

/**
 * Pure payment-state evaluator: given a request, an observation, and the current
 * time, compute the status. No I/O — deterministic and unit-testable.
 */
export function evaluatePayment(
  request: PaymentRequest,
  observation: PaymentObservation,
  now: number = Date.now(),
): PaymentState {
  const state = (status: PaymentStatus): PaymentState => ({
    status,
    request,
    observation,
    terminal: TERMINAL.has(status),
  });

  const expected = request.amountBaseUnits;
  const confirmed = observation.confirmedBaseUnits;
  const received = observation.receivedBaseUnits;

  // Fully confirmed outcomes take precedence over expiry.
  if (confirmed >= expected) {
    return state(confirmed > expected ? "overpaid" : "paid");
  }
  // Expiry only applies before the amount is confirmed.
  if (now >= request.expiresAt) {
    // If some confirmed funds arrived but not enough, it's an underpayment.
    if (confirmed > 0n) return state("underpaid");
    return state("expired");
  }
  if (!observation.seen || received === 0n) {
    return state("pending");
  }
  if (observation.confirmations < request.requiredConfirmations) {
    // Seen in mempool / not yet deep enough.
    return state(observation.confirmations === 0 ? "detected" : "confirming");
  }
  // Enough confirmations but not enough value.
  return state("underpaid");
}

/** Build a webhook-ready event object from a payment state. */
export function toPaymentEvent(state: PaymentState, now: number = Date.now()): PaymentEvent {
  return {
    type: "payment_status",
    paymentId: state.request.id,
    status: state.status,
    address: state.request.address,
    expectedBaseUnits: state.request.amountBaseUnits.toString(),
    receivedBaseUnits: state.observation.receivedBaseUnits.toString(),
    confirmations: state.observation.confirmations,
    requiredConfirmations: state.request.requiredConfirmations,
    at: now,
  };
}

/**
 * Observe a payment via wallet RPC. Sums unspent outputs paid to the request's
 * address. Assumes a unique, unspent-until-settled address (the recommended
 * pattern). For swept/spent funds, track receipts in your own database.
 */
export async function observePayment(
  client: KingPepeClient,
  request: PaymentRequest,
): Promise<PaymentObservation> {
  const utxos = await client.listUnspent(0, 9_999_999, [request.address]);
  let received = 0n;
  let confirmed = 0n;
  let maxConf = 0;
  let seen = false;
  for (const u of utxos) {
    seen = true;
    const base = rpcAmountToBaseUnits(u.amount);
    received += base;
    if (u.confirmations >= request.requiredConfirmations) confirmed += base;
    if (u.confirmations > maxConf) maxConf = u.confirmations;
  }
  return { receivedBaseUnits: received, confirmedBaseUnits: confirmed, confirmations: maxConf, seen };
}

/** Convenience: observe + evaluate in one call. */
export async function checkPayment(
  client: KingPepeClient,
  request: PaymentRequest,
  now: number = Date.now(),
): Promise<PaymentState> {
  const observation = await observePayment(client, request);
  return evaluatePayment(request, observation, now);
}
